File: 2024/01/go/part2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func solvePart2(in string) string {
	// Split input into lines
	lines := strings.Split(strings.TrimSpace(in), "\n")

	// Extract left and right numbers
	var leftNums []int
	rightFreq := make(map[int]int)

	for _, line := range lines {
		var left, right int
		n, _ := fmt.Sscanf(line, "%d %d", &left, &right)
		if n == 2 {
			// Store left number in slice
			leftNums = append(leftNums, left)
			// Count frequency of right numbers
			rightFreq[right]++
		}
	}

	// Calculate similarity score
	similarityScore := 0
	for _, leftNum := range leftNums {
		// For each number in left list, multiply by its frequency in right list
		similarityScore += leftNum * rightFreq[leftNum]
	}

	return strconv.Itoa(similarityScore)
}

File: 2024/01/java/src/main/java/net/advw/aoc/Part2.java
package net.advw.aoc;

public class Part2 {
    public static String solve(String input) {

        // Parse input string into arrays
        String[] lines = input.trim().split("\n");
        int[] leftNumbers = new int[lines.length];
        int[] rightNumbers = new int[lines.length];

        // Fill arrays from input
        for (int i = 0; i < lines.length; i++) {
            String[] parts = lines[i].trim().split("\\s+");
            leftNumbers[i] = Integer.parseInt(parts[0]);
            rightNumbers[i] = Integer.parseInt(parts[1]);
        }

        // Calculate similarity score
        int totalScore = 0;

        // For each number in left list
        for (int leftNum : leftNumbers) {
            // Count occurrences in right list
            int occurrences = 0;
            for (int rightNum : rightNumbers) {
                if (rightNum == leftNum) {
                    occurrences++;
                }
            }
            // Add to total score (number * occurrences)
            totalScore += leftNum * occurrences;
        }

        return String.valueOf(totalScore);
    }
}
